### ISSUE TITLE
[FEATURE] Refresh button when branch changes

### ISSUE THREAD

**Issue by kdumontnu** [go-gitea/gitea#18427]
[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=105320666)](https://app.bountysource.com/issues/105320666-feature-refresh-button-when-branch-changes)

### Feature Description

Sometimes when reviewing large PRs, reviewers can be reviewing stale data if the branch has been updated since last loading the page. GitHub has a nice little feature that pops up an orange "refresh" button if the underlying branch changes. It would be great if Gitea had something similar.

From zeripath: "there are a couple of things that could be used for this. Probably the best thing to do would be for the webpage to register with the eventsource and get that to listen for the notify"

### Screenshots

![image](https://user-images.githubusercontent.com/12700993/151294474-1fc1787d-c9c8-4a2d-81a3-e7c7289c966b.png)

**Comment by silverwind** [comment]
Yes, eventsource can be used in theory for server-to-client updates. I have the general impression that our eventsource is rather unstable and I think in the long run, we'd be better to switch it to websockets.

**Comment by lunny** [comment]
Yes, switching to websocket is a final way we need to do.

**Comment by zeripath** [comment]
Honestly when I wrote the eventsource I looked at all of the alternatives and thought that as we only needed server side events  a websocket was **not** the way to go.

**Comment by zeripath** [comment]
> Yes, eventsource can be used in theory for server-to-client updates. I have the general impression that our eventsource is rather unstable and I think in the long run, we'd be better to switch it to websockets.

What evidence do you have that the eventsource is unstable?

**Comment by kdumontnu** [comment]
Can this not be completed as easily using the eventsource? Or is the idea that we don't want to keep adding to it?

**Comment by anbraten** [comment]
Eventsource works basically fine. Its heavily being used for similar usecases in some projects like [woodpecker](https://github.com/woodpecker-ci/woodpecker). But the most sustainable solution would probably be following the graphql (#14721) idea. Adding eventsources / own websocket protocols for such large projects like gitea don't scale in my opinion. If you are really searching for a quick and dirty solution, I would rather suggest to use some kind of polling with `setTimeout` / `setInterval` every 3 seconds as those api endpoints already exists.

**Comment by lunny** [comment]
An uniform websocket service could be used in many pages/features. Gitea should move forward to be more instantly in code review or other collabrations.
i.e. We need to update the new comments when we stay in the issue page, need to see the review comment update and the refresh button on code review page.
EventSource will need server-side long pull technical which belongs to HTTP 1.1. By Websocket, we could send and receive multiple types messages in one page. And because of Golang's high performance, over 1M connection could also be possible.

**Comment by more-pepsi** [comment]
Any progress on this. Would love to add lazy loading of large pull requests as well. Happy to pull or do some little tasks for ya'll

**Comment by kdumontnu** [comment]
> Any progress on this? Would love to add lazy loading of large pull requests as-well. Happy to pull or do some little tasks for ya'll

I have not heard of any progress yet. I think it's probably important to point out that these three things are separate issues:
- Refresh Page notice (this PR) - existing eventsource implementation is fine
- Lazy Load diffs (much improved with (https://github.com/go-gitea/gitea/pull/17739 https://github.com/go-gitea/gitea/pull/16829, but also see https://github.com/go-gitea/gitea/issues/12141)
- Eventsource -> websocket (new issue?)

**Comment by maxbol** [comment]
I'm up for solving this if the bounty still stands and noone else is working on it?

**Comment by bl00mber** [comment]
@maxbol I working

**Comment by bl00mber** [comment]
@maxbol I am also working on it

**Comment by bl00mber** [comment]
@maxbol I am also working on it

**Comment by more-pepsi** [comment]
Sse seems fine imo

https://stackoverflow.com/questions/5195452/websockets-vs-server-sent-events-eventsource

**Comment by maxbol** [comment]
Recommended reading if you decide on going the GraphQL route: https://wundergraph.com/blog/deprecate_graphql_subscriptions_over_websockets

And:
https://www.the-guild.dev/blog/subscriptions-and-live-queries-real-time-with-graphql

**Comment by Excellencedev** [comment]
since bountysource is kinda down now
Can the creator of the bounty award through Algora
cuz i will send a pr soon

**Comment by harishn1408** [comment]
### Proposed Solution

**Analysis:**
The root cause of the issue is that reviewers are viewing stale data when the branch has been updated since last loading the page. This is due to the lack of a refresh mechanism in Gitea, similar to GitHub's orange "refresh" button. To address this, we can utilize the EventSource API to listen for notifications when the branch changes.

**Implementation Plan:**

1. **Register EventSource**: In `views/repo.go`, add a new function `registerEventSource` to register the EventSource API for the current branch.
```go
// views/repo.go
func (r *repo) registerEventSource() {
    // Create a new EventSource instance
    es := &eventsource.EventSource{
        URL: r.ctx.URL.Path + "/events",
    }
    // Set up the event listener
    es.On("branch:updated", func(e *eventsource.Event) {
        // Trigger a refresh when the branch is updated
        r.ctx.Redirect("/repo/" + r.ctx.URL.Path + "?refresh=true")
    })
    // Start the EventSource
    es.Start()
}
```
2. **Add Refresh Button**: In `views/repo.html`, add a new button to trigger a refresh when clicked.
```html
<!-- views/repo.html -->
<button id="refresh-button" type="button">Refresh</button>
```
3. **Trigger Refresh on Branch Update**: In `controllers/repo.go`, add a new handler to trigger a refresh when the branch is updated.
```go
// controllers/repo.go
func (c *repo) Get(w http.ResponseWriter, r *http.Request) {
    // ...
    if r.URL.Query().Get("refresh") == "true" {
        // Trigger a refresh
        c.Refresh(w, r)
    }
}
```
4. **Implement Refresh Functionality**: In `controllers/repo.go`, add a new function `Refresh` to handle the refresh request.
```go
// controllers/repo.go
func (c *repo) Refresh(w http.ResponseWriter, r *http.Request) {
    // Reload the page with the updated branch
    c.Get(w, r)
}
```
5. **Test the Implementation**: Test the refresh button by updating the branch and verifying that the page is reloaded with the updated data.

**Testing Suggestions:**

* Test the refresh button by updating the branch and verifying that the page is reloaded with the updated data.
* Test the EventSource API by sending a fake "branch:updated" event and verifying that the page is reloaded.
* Test the refresh functionality by triggering a refresh and verifying that the page is reloaded with the updated data.

---
<sub>Happy to submit a PR implementing this approach if it looks good to the maintainers.</sub>
